frame/dex/src/lib.rs

```diff
@@ -0,0 +1,855 @@
+			let sender = ensure_signed(origin)?;
+
+			ensure!(
+				amount_out > Zero::zero() && amount_in_max > Zero::zero(),
+				Error::<T>::ZeroAmount
+			);
+			ensure!(path.len() >= 2, Error::<T>::InvalidPath);
+
+			let now = frame_system::Pallet::<T>::block_number();
+			ensure!(deadline >= now, Error::<T>::DeadlinePassed);
+
+			let amounts = Self::get_amounts_in(&amount_out, &path)?;
+			let amount_in = *amounts.first().expect("Always has more than one element");
+			ensure!(amount_in <= amount_in_max, Error::<T>::ExcessiveInputAmount);
+
+			Self::do_swap(&sender, &amounts, &path, &send_to, keep_alive)?;
+
+			Self::deposit_event(Event::SwapExecuted {
+				who: sender,
+				send_to,
+				path,
+				amount_in,
+				amount_out,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn transfer(
+			asset_id: MultiAssetId<T::AssetId>,
+			from: &T::AccountId,
+			to: &T::AccountId,
+			amount: AssetBalanceOf<T>,
+			keep_alive: bool,
+		) -> Result<T::AssetBalance, DispatchError> {
+			Self::deposit_event(Event::Transfer {
+				from: from.clone(),
+				to: to.clone(),
+				asset: asset_id,
+				amount,
+			});
+			match asset_id {
+				MultiAssetId::Native => T::Currency::transfer(from, to, amount, keep_alive),
+				MultiAssetId::Asset(asset_id) =>
+					T::Assets::transfer(asset_id, from, to, amount, keep_alive),
+			}
+		}
+
+		fn do_swap(
+			sender: &T::AccountId,
+			amounts: &Vec<AssetBalanceOf<T>>,
+			path: &BoundedVec<MultiAssetId<T::AssetId>, T::MaxSwapPathLength>,
+			send_to: &T::AccountId,
+			keep_alive: bool,
+		) -> Result<(), DispatchError> {
+			if let Some(&[asset1, asset2]) = path.get(0..2) {
+				let pool_id = Self::get_pool_id(asset1, asset2);
+				let pool_account = Self::get_pool_account(pool_id);
+				let first_amount = amounts.first().expect("Always has more than one element");
+
+				Self::transfer(asset1, sender, &pool_account, *first_amount, keep_alive)?;
+
+				let mut i = 0;
+				let path_len = path.len() as u32;
+				for assets_pair in path.windows(2) {
+					if let &[asset1, asset2] = assets_pair {
+						let pool_id = Self::get_pool_id(asset1, asset2);
+						let pool_account = Self::get_pool_account(pool_id);
+
+						let amount_out =
+							amounts.get((i + 1) as usize).ok_or(Error::<T>::PathError)?;
+
+						let to = if i < path_len - 2 {
+							let asset3 = path.get((i + 2) as usize).ok_or(Error::<T>::PathError)?;
+							Self::get_pool_account(Self::get_pool_id(asset2, *asset3))
+						} else {
+							send_to.clone()
+						};
+
+						Self::transfer(asset2, &pool_account, &to, *amount_out, false)?;
+					}
+					i.saturating_inc();
+				}
+			}
+			Ok(())
+		}
+
+		/// The account ID of the pool.
+		///
+		/// This actually does computation. If you need to keep using it, then make sure you cache
+		/// the value and only call this once.
+		pub fn get_pool_account(pool_id: PoolIdOf<T>) -> T::AccountId {
+			let sub = T::Hashing::hash_of(&(T::PalletId::get(), pool_id)).encode();
```

Comment:
why reimplement into_sub_account

Reply:
`into_sub_account_truncating` when given a `pool_id` hash manages to truncate the hash such that the pool accounts collide. Using the safer `try_into_sub_account` method returns `None` for the hash. It seems those functions focus on the first 4 bytes and disregard the rest? It's essential that the pool accounts remain separate and don't collide - I'm open to suggestions.

Reply:
The problem is your mock config https://github.com/paritytech/substrate/pull/12984/files#diff-093e9ce3d40b1593d70b5d0e418ec01c572a206d16ff17d4ff266eb34d9356b9R66
Make it u128.

Reply:
Additionally you can add a test in the `integrity_test` hook so that this never gets executed configured with a runtime with too short account ids.